src/java.base/share/classes/java/util/Collections.java

```diff
@@ -5798,4 +5798,61 @@ public boolean removeIf(Predicate<? super E> filter) {
         @Override
         public Stream<E> parallelStream()   {return q.parallelStream();}
     }
+
+    /**
+     * Calculate initial capacity for HashMap based classes, from expected size.
+     *
+     * @param expectedSize expected size
+     * @return initial capacity for HashMap based classes.
+     * @since 19
+     */
+    private static int calculateHashMapCapacity(int expectedSize) {
+        if (expectedSize >= 1610612736) {
+            return Integer.MAX_VALUE;
+        }
+        return (expectedSize + (expectedSize + 2) / 3);
+    }
+
+    /**
+     * Create new HashMap from expected size.
+     *
+     * @param expectedSize expected size
+     * @param <K>          the type of keys maintained by this map
+     * @param <V>          the type of mapped values
+     * @return initial capacity for HashMap based classes.
+     * @throws IllegalArgumentException if the initial capacity is negative.
+     * @since 19
+     */
+    public static <K, V> HashMap<K, V> newHashMap(int expectedSize) {
```

Comment:
Stuart [in a JBS comment](https://bugs.openjdk.java.net/browse/JDK-8186958?focusedCommentId=14483078&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14483078) recommends keeping these factories to the respective classes instead of `Collections`. In existing code, such factories stay in the respective classes, including `ConcurrentHashMap.newKeySet(int)`.

Reply:
good but I have no account to comment on your jira system.

Reply:
still I think it better to put these functions to Collections as they can share a same calculation function.

Reply:
> still I think it better to put these functions to Collections as they can share a same calculation function.

You can just keep these methods in `Collections` but make them package-private, so the moved factories can call. (Note that the methods may have to go to somewhere else in `java.util` if `Collections` cannot be loaded too early during startup)

> good but I have no account to comment on your jira system.

I think when the patch looks good, stuart will file one CSR for you, or you can write a CSR draft for him to file.

Reply:
> Note that the methods may have to go to somewhere else in `java.util` if `Collections` cannot be loaded too early during startup, as HashMap is loaded very early

I'm afraid you are correct...

Reply:
@liach done.